Security Team Goals (#1585)
sourcegraph/about: handbook/engineering/security/index.md

## Goals

### Long term

> Every organization, is confident they can trust Sourcegraph with the security of their Source code.

Reviewer: ```suggestion
Every organization is confident they can trust Sourcegraph cloud to securely store and access their source code.
```

### Medium term

The goals below lie on the path to our long-term goals, and represent the next 3-6 months of work.  These goals will be delivered serially, but may be worked on in parallel.

> **We share our private [infrastructure repository](http://github.com/sourcegraph/infrastructure) on Sourcegraph cloud**

Reviewer: ```suggestion
**We store our private [infrastructure repository](http://github.com/sourcegraph/infrastructure) on Sourcegraph cloud**
```
Assuming you mean store or index here, since we still want it to be private

Author: Updated to host in latest.

>   - Problem and rationale: While Sourcegraph cloud is in an good, and improvine security state, we currently do not have the right level of visibility. By starting with a private repository that contains no secrets, we can build and deploy the visibility technologies and processes that allow us to build a holistic view of the environment.  This serves as a proxy for small company repository use.

Reviewer: ```suggestion
  - Problem and rationale: While Sourcegraph cloud is in a good state in regards to security, we currently do not have the right level of visibility. We need more visibility into our threat surface before we feel confident putting private code in our cloud service. We can use our own comfort level as a proxy for how other small companies would feel when asked to put private code on cloud. We want to be able to say that we have good visibility into our threat surface, and we want to be confident that security issues will be detected and fixed before they become problematic.
```

Reworded this to hopefully make it flow better. Fundamentally, we want to get cloud into a state where we're confident in putting our own private code out there, and where we're confident that we'll know about vulnerabilities very quickly.

  - Planned work:
    - Run ongoing vulnerability scans of the containers including vulnerability remediation
    - Deploy centralized security logging
    - Ship container logs to the logging destination
    - Begin documenting policies

>   - Definition of success: Frequent container vulnerability scans are being run, and stored for future a audit. Fixes to the containers or their underlying service configurations have made it into the product release cycle.  Logs of new container deployments are available for search in a centralized logging tool.

Reviewer 2: ```suggestion
  - Definition of success: Frequent container vulnerability scans are being run, and stored for a future audit. Fixes to the containers or their underlying service configurations have made it into the product release cycle.  Logs of new container deployments are available for search in a centralized logging tool.
```

> **We share a private repository containing encrypted secrets**

Reviewer 2: This isn't clear. Where are we sharing and with whom?

Author: I'm trying to capture that Sourcegraph hosts these repositories.  For now, I've changed We share to We host, but I'm open to better phrasing.

Reviewer 2: Host is much better, thanks.

>   - Problem and rationale: Customers will have repositories containing encrypted secrets, using an [ansible vault](https://docs.ansible.com/ansible/latest/user_guide/vault.html) or gpg encrypted files. Trusting these repositories to external systems requires the added degree of care and concern, expected in security conscious small and medium enterprise (SME). To support those needs, not only is enhanced visibility into our own systems required on our side, it will be contractually obligated, or at a minimum part of security questionnaires.

Reviewer 2: This wording makes it seam like we will be encrypting customer secrets with either of these solutions. Is that the case?   Could we phrase this as a problem to be solved rather than a solution?

Author: Where did you view this as Sourcegraph encrypting customer secrets? I can definitely see changing the first sentence - and have pushed up a change that I believe addresses that.

Reviewer 2: > Customers will have repositories containing encrypted secrets, using an [ansible vault](https://docs.ansible.com/ansible/latest/user_guide/vault.html) or gpg encrypted files.

"Will have" makes it seam like we're going to do this.

Author: Also updated in that same push. Text now shows as *Customers repositories can contain encrypted secrets, such as an [ansible vault](https://docs.ansible.com/ansible/latest/user_guide/vault.html) or gpg encrypted files.*

>   - Planned work:

Reviewer: I think we need some sort of RBAC for this. At the very least, we need to reduce the scope of admins from site admins to org admins, so that admins can manage their users without being able to create users in other orgs.

Reviewer 2: @ElizabethStirling: Any substantial changes to our authz and authn systems would currently be driven by the Cloud team as part of their backend infrastructure responsibilities. Please sync with @unknwon to make sure you're on the same page.

Reviewer: I did some investigation into this, and I think I'm going to walk back my thoughts on RBAC for now. I'm seeing a way that our current state, or at least a state very close to our current state, could be fine for getting cloud to a state where we're comfortable shipping it to customers.

    - Run ongoing vulnerability scans of underlying workloads including vulnerability remediation
    - Deploy host-based intrusion detection (HIDS)
    - Centralize cloud and workload logs in a single location
    - Add support for audit and access logging
    - Normalize security and application logs
    - Create dashboards and alerts for events of interest
    - Extend policy documentation
  - Definition of success: We deploy a private repository containing encrypted secrets to Sourcegraph cloud, complete with a baseline for security monitoring. We are able to validate access to underlying hosts and the application in centralized logging.

**Host a capture the flag for Sourcegraph.com**

>   - Problem and rationale: Running a world class service capable of hosting private repositories is about more than application security. As examples, we need visibility into data loss, processes around incident response need documentation, and defensive measures for the internal actor need to be expanded upon. The proxy enhances the addressed security concerns that a medium enterprise would require.

Reviewer: ```suggestion
  - Problem and rationale: Running a world class service capable of hosting private repositories is about more than application security. As examples, we need visibility into data loss, processes around incident response need documentation, and expanded defensive measures against internal threat actors. The proxy enhances the addressed security concerns that a medium enterprise would require.
```

Author: Modified your text to and reworked. Perhaps this is better?

   
### Learning

> After the each iteration, we hold a [retrospective](https://about.sourcegraph.com/retrospectives). We try to understand the degree to which we achieved the goals we communicaed at the beginning.  We idenitfy what went well and what our opportunities for imporvemen, and actively choose one of the things we've learned, and improve it.

Reviewer: ```suggestion
After the each iteration, we hold a [retrospective](https://about.sourcegraph.com/retrospectives). We try to understand the degree to which we achieved the goals we communicated at the beginning.  We identify what went well and what our opportunities for improvement are, then actively choose one of the things we've learned to integrate into our team's ethos.
```
